subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,234 @@
+#include <hal/nrf_power.h>
+#include <debug/ppi_trace.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(cpu_load, CONFIG_CPU_LOAD_LOG_LEVEL);
+
+/* Convert event address to associated publish register */
+#define PUBLISH_ADDR(evt) (volatile u32_t *)(evt + 0x80)
+
+/* Define to please compiler when periodic logging is disabled. */
+#ifndef CONFIG_CPU_LOAD_LOG_INTERVAL
+#define CONFIG_CPU_LOAD_LOG_INTERVAL 0
+#endif
+
+static nrfx_timer_t timer = NRFX_TIMER_INSTANCE(CONFIG_CPU_LOAD_TIMER_INSTANCE);
+static bool ready;
+static struct k_delayed_work cpu_load_log;
+
+/** @brief Allocate (D)PPI channel. */
+static nrfx_err_t ppi_alloc(u8_t *ch, u32_t evt)
+{
+	nrfx_err_t err;
+#ifdef DPPI_PRESENT
+	if (*PUBLISH_ADDR(evt) != 0) {
+		/* Use mask of one of subscribe registers in the system,
+		 * assuming that all subscribe registers has the same mask for
+		 * channel id.
+		 */
+		*ch = *PUBLISH_ADDR(evt) & DPPIC_SUBSCRIBE_CHG_EN_CHIDX_Msk;
+		err = NRFX_SUCCESS;
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+static void cpu_load_log_fn(struct k_work *item)
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load/1000;
+	u32_t fraction = load - 1000*percent;
+
+	cpu_load_reset();
+	LOG_INF("Load:%d,%03d%%", percent, fraction);
+	k_delayed_work_submit(&cpu_load_log,
+			CONFIG_CPU_LOAD_LOG_INTERVAL);
+}
+
+static int cpu_load_log_init(void)
+{
+	k_delayed_work_init(&cpu_load_log, cpu_load_log_fn);
+	return k_delayed_work_submit(&cpu_load_log,
+			      CONFIG_CPU_LOAD_LOG_INTERVAL);
+}
+
+void timer_handler(nrf_timer_event_t event_type, void *context)
+{
+	/*empty*/
+}
+
+int cpu_load_init(void)
+{
+	u8_t ch_sleep;
+	u8_t ch_wakeup;
+	u8_t ch_tick = 0;
+	nrfx_err_t err;
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	int ret = 0;
+
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		/* It's assumed that RTC1 is driving system clock. */
+		config.mode = NRF_TIMER_MODE_COUNTER;
+		err = ppi_alloc(&ch_tick,
+		       nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK));
+		if (err != NRFX_SUCCESS) {
+			return -ENODEV;
+		}
+		nrfx_gppi_channel_endpoints_setup(ch_tick,
+		     nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK),
+		     nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_COUNT));
+		nrf_rtc_event_enable(NRF_RTC1, NRF_RTC_INT_TICK_MASK);
+	}
+
+	if (ready) {
+		return 0;
+	}
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
+	}
+
+	err = ppi_alloc(&ch_sleep,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPENTER));
+	if (err != NRFX_SUCCESS) {
+		return -ENODEV;
+	}
+
+	err = ppi_alloc(&ch_wakeup,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		/* ppi free */
```

Comment:
Was that (freeing of the allocated channels) planned to be added?

Reply:
added ppi cleanup which frees channels. However, with DPPI it is not that simple as cpu_load supports a case when event is already allocated and then it just subscribes to it. I added it as an option (and extended test to check that).

I also had to take into account this when enabling ppi channels. "shared" channels are not enabled.